internal/app/public.go

```diff
@@ -1,109 +1,145 @@
 		}
 
-		dssid, err := id.DatasetSchemaIDFrom(inp.DatasetSchemaID)
+		output, err := controller.CreateUser(c.Request().Context(), inp)
 		if err != nil {
-			return &echo.HTTPError{Code: http.StatusNotFound, Message: err1.ErrNotFound}
+			return err
 		}
-		if inp.Author == "" {
-			return &echo.HTTPError{Code: http.StatusBadRequest, Message: errors.New("require author value")}
+
+		return c.JSON(http.StatusOK, output)
+	})
+
+	r.GET("/published/:name", func(c echo.Context) error {
+		name := c.Param("string")
+		if name == "" {
+			return echo.ErrNotFound
 		}
-		if inp.Content == "" {
-			return &echo.HTTPError{Code: http.StatusBadRequest, Message: errors.New("require content value")}
+
+		res, err := publishedController.Metadata(c.Request().Context(), name)
+		if err != nil {
+			return err
 		}
-		interactor := interactor.NewDataset(repos, gateways)
-		dss, ds, err := interactor.AddDynamicDataset(c.Request().Context(), interfaces.AddDynamicDatasetParam{
-			SchemaId: dssid,
-			Author:   inp.Author,
-			Content:  inp.Content,
-			Lat:      inp.Lat,
-			Lng:      inp.Lng,
-			Target:   inp.Target,
-		})
 
+		return c.JSON(http.StatusOK, res)
+	})
+
+	r.GET("/published_data/:name", func(c echo.Context) error {
+		name := c.Param("string")
+		if name == "" {
+			return echo.ErrNotFound
+		}
+
+		r, err := publishedController.Data(c.Request().Context(), name)
 		if err != nil {
-			if errors.Is(err1.ErrNotFound, err) {
-				return &echo.HTTPError{Code: http.StatusNotFound, Message: err}
-			}
-			return &echo.HTTPError{Code: http.StatusInternalServerError, Message: err}
+			return err
 		}
-		response := make(map[string]interface{})
-		response["id"] = ds.ID().String()
-		for _, f := range dss.Fields() {
-			response[f.Name()] = toResponseValue(ds.Field(f.ID()).Value())
+
+		return c.Stream(http.StatusOK, "application/json", r)
+	})
+}
+
+func publishedRoute(
+	ec *echo.Echo,
+	r *echo.Group,
+	conf *Config,
+	repos *repo.Container,
+	gateways *gateway.Container,
+) {
+	var i interfaces.Published
+	if conf.Published.IndexURL == nil || conf.Published.IndexURL.String() == "" {
+		html, err := os.ReadFile("web/published.html")
+		if err == nil {
+			i = interactor.NewPublished(repos.Project, gateways.File, string(html))
+		} else {
+			i = interactor.NewPublished(repos.Project, gateways.File, "")
 		}
-		return c.JSON(http.StatusOK, response)
+	} else {
+		i = interactor.NewPublishedWithURL(repos.Project, gateways.File, conf.Published.IndexURL)
+	}
+	contr := http1.NewPublishedController(i)
+
+	key := struct{}{}
+	auth := middleware.BasicAuthWithConfig(middleware.BasicAuthConfig{
+		Validator: func(user string, password string, c echo.Context) (bool, error) {
+			md, ok := c.Request().Context().Value(key).(interfaces.ProjectPublishedMetadata)
+			if !ok {
+				return true, echo.ErrNotFound
+			}
+			return !md.IsBasicAuthActive || user == md.BasicAuthUsername && password == md.BasicAuthPassword, nil
```

Comment:
This is not a critical problem but not to be guessed values from the time to compare, it's recommended to use [ConstantTimeCompare](https://echo.labstack.com/middleware/basic-auth/) to compare values.

Reply:
Great. Thank you.